sympy/physics/mechanics/joint.py

```diff
@@ -0,0 +1,681 @@
+           'SphericalJoint', 'PlanarJoint']
+
+
+class Joint(object):
+    """Abstract Base class for all specific joints.
+
+    A Joint connects two bodies (a parent and child) by adding different degrees
+    of freedom to child with respect to parent.
+    This is the base class for all specific joints and holds all common methods
+    acting as an interface for all joints. Custom joint can be created by
+    creating a subclass of this class and overriding 'apply_joint()' method to
+    add the dynamics of the joint.
+
+    Parameters
+    ----------
+    name: String
+        Name of the joint which makes it unique. Should be different from other
+        joints.
+    parent: Body
+        Instance of Body class serving as the parent in the joint.
+    child: Body
+        Instance of Body class serving as the child in the joint.
+    parent_point_pos: 3 Tuple, Optional
+        Defines the joint's point where the parent will be connected to child.
+        3 Tuple defines the values of x, y and z directions w.r.t parent's
+        frame. If it is not supplied, center of mass is added as default.
+    child_point_pos: 3 Tuple, Optional
+        Defines the joint's point where the child will be connected to parent.
+        3 Tuple defines the values of x, y and z directions w.r.t child's frame.
+        If it is not supplied, center of mass is added as default.
+    parent_axis: Vector, Optional
+        Defines the orientation as a vector which must be aligned with child's
+        axis before adding joint. If it is not passed, default is x axis in
+        parent's frame.
+    child_axis: Vector, Optional
+        Defines the orientation as a vector which must be aligned with parent's
+        axis before adding joint. If it is not passed, default is x axis in
+        child's frame.
+    """
+
+    def __init__(self, name, parent, child, parent_point_pos=None,
+                 child_point_pos=None):
+        self.name = name
+        self.parent = parent
+        self.child = child
+        self.coordinates = []
+        self.speeds = []
+        self.kds = []
+
+        if parent_point_pos is None:
+            parent_point_pos = (0, 0, 0)  # Parent's Center of mass
+
+        if child_point_pos is None:
+            child_point_pos = (0, 0, 0)  # Child's Center of mass
+
+        self._parent_joint_location = convert_tuple_to_vector(
+            parent.frame, parent_point_pos)
+        self._child_joint_location = convert_tuple_to_vector(
+            child.frame, child_point_pos)
+
+        self._locate_joint_point()
+        self.apply_joint()
+
+    def _locate_joint_point(self):
+        self.parent_joint_point = self.parent.masscenter.locatenew(
+            self.name + '_parent_joint',
+            self._parent_joint_location)
+        self.child_joint_point = self.child.masscenter.locatenew(
+            self.name + '_child_joint',
+            self._child_joint_location)
+
+    def _align_axes(self, parent_axis, child_axis):
```

Comment:
@moorepants While going through examples and looking at this work for making a better and complete PR, this method is really bothering me.

Why do we need this function?

Reply:
I think this class ensures that a point and vector fixed on each body are made coincident and collinear, respectively.

Reply:
But I'm not sure exactly, only tests and examples can show us the intent.